pkg/gardenlet/bootstrap/certificate/certificate_rotation_test.go

```diff
@@ -0,0 +1,342 @@
+				validity := 20 * time.Second
+				cert := generateCertificate(validity)
+				gardenKubeconfigWithValidClientCert = fmt.Sprintf(baseKubeconfig, utils.EncodeBase64(cert.CertificatePEM), utils.EncodeBase64(cert.PrivateKeyPEM))
+				gardenKubeconfigWithInValidClientCert = fmt.Sprintf(baseKubeconfig, "bm90LXZhbGlk", utils.EncodeBase64(cert.PrivateKeyPEM))
+			})
+
+			It("should not return an error", func() {
+				// mock existing secret with garden kubeconfig
+				updateKubeconfigSecret(k8sClient, gardenClientConnection, gardenKubeconfigWithValidClientCert)
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				cert, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(cert).ToNot(BeNil())
+			})
+
+			It("should return an error - kubeconfig secret does not exist", func() {
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should return an error - secret does not contain a kubeconfig", func() {
+				// mock existing secret with garden kubeconfig
+				kubeconfigSecret := &corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      gardenClientConnection.KubeconfigSecret.Name,
+						Namespace: gardenClientConnection.KubeconfigSecret.Namespace,
+					},
+				}
+				err := k8sClient.Create(context.TODO(), kubeconfigSecret)
+				Expect(err).ToNot(HaveOccurred())
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err = getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should return an error - kubeconfig client cert is invalid", func() {
+				// mock existing secret with garden kubeconfig
+				updateKubeconfigSecret(k8sClient, gardenClientConnection, gardenKubeconfigWithInValidClientCert)
+
+				fakeClientMap := fakeclientmap.NewClientMap().
+					AddClient(keys.ForGarden(), k8sGardenClient).
+					AddClient(keys.ForSeed(seed), k8sGardenClient)
+
+				_, err := getCurrentCertificate(context.TODO(), gardenClientConnection, fakeClientMap, seedName)
+				Expect(err).To(HaveOccurred())
+			})
+		})
+
+		Describe("#waitForCertificateRotation", func() {
+			var (
+				fakeClientMap  clientmap.ClientMap
+				mockClient     *mockclient.MockClient
+				ctx            = context.TODO()
+				testKubeconfig string
+			)
+
+			BeforeEach(func() {
+				mockClient = mockclient.NewMockClient(ctrl)
+
+				validity := 10 * time.Second
+				cert := generateCertificate(validity)
```

Comment:
Please do not generate certificates in the unit test and rather load an existing certificate in the tests (either define a string const containing the certificate or load it from some test resource file).
This might take long and we should probably avoid such tests, if it is easy like here.

Reply:
I do not think that this is critical here. Unlike at runtime where we want to avoid high cpu usage, it does not matter when we execute this function 3 times during a test suite execution. 
I would leave it like this to avoid additional effort (functionality wise it is doing what it should).

More opinions welcome.

Reply:
Hmm, ok